### PR TITLE
Make THCTensor_varInnermostDim numerically stable

### PR DESCRIPTION
**Pull request by zou3519**
Addressing #3402 and #2470. #3410 has some relevant information about how this works.

Uses Welford's variance algorithm for numeric stability. 

### Test Plan
`test/run_test.sh`
Added unit tests for variance on cuda tensors. One in particular tests large, uneven CUDA inputs to make sure the code computes all the variances (as opposed to only some of them).

Check numeric stability:
```
import torch
tensor = torch.FloatTensor([2281.5, 2281.25]).cuda()
tensor.var(0)  # Used to be 0. Now is 0.03125
```

### CODE REVIEW

**Comment by ngimel** [comment]
Looks good, unfortunately to make halfs usable all the intermediate computations have to be in float32, so you have to use Real and AccReal templated types (it is done, e.g., in all the reduction kernels). It can be left for future PR's, though, because in any case it is better than what is done now :-)

**Comment by zou3519** [comment]
That makes sense, I can add that in on this PR.

File: test/test_torch.py [hunk 1]
```diff
@@ -3605,6 +3605,32 @@ def test_unbiased(self):
         self.assertEqual(tensor.std(), tensor.std(unbiased=True))
         self.assertEqual(tensor.std(unbiased=False), tensor.std(0, unbiased=False)[0])
 
+    @unittest.skipIf(not torch.cuda.is_available(), 'no CUDA')
+    def test_unbiased_cuda(self):
+        tensor = torch.randn(100).cuda()
+        self.assertEqual(tensor.var(0), tensor.var(0, unbiased=True))
```
Comment (colesbury) [inline test/test_torch.py]: This looks like it's just testing that `unbiased=True` is the default. That's not very useful.

Test that the outputs match the CPU functions or that it matches some known value.

Since you are improving numerical precision, add a test for values that you know would cause unacceptable numerical errors using the previous method.

**Comment by ezyang** [comment]
Since the great ATen merge landed, the changes to THC have to be moved over to the ones in aten. Shout if you need help.

**Review by colesbury — COMMENTED** [review]
Nice work!

Can you post before and after numbers for calls to `var()` with a few different sizes? I find %timeit in ipython usually sufficient for this, e.g.:

```
%timeit tensor.var(...); torch.cuda.synchronize()
```

**Comment by zou3519** [comment]
Here are some numbers for various calls to `var(dim)` (thanks for the %timeit tip!) https://gist.github.com/zou3519/c301985417b0f22f6c34ff814d583008

In general after the numeric stability changes the calls are slower. The largest difference (new_time / old_time) is around 1.7

**Comment by colesbury** [comment]
Nice, this looks good to me and the perf hit seem worth the numerical stability.

**Comment by ngimel** [comment]
@zou3519, can you do the same thing for varOuterDim? https://github.com/pytorch/pytorch/blob/master/aten/src/THC/THCTensorMathReduce.cuh#L309. 
It's much simpler as there is no reduction across the threads, so just repeating what you did for serial case will suffice.

**Comment by zou3519** [comment]
Yup, I'll take a look at that

**Comment by colesbury** [comment]
@zou3519: @soumith told me that we only care about compute capability 3.0+ (Kepler+) now. I think all the reductions here are within a warp. You can probably use the [warp shuffle](http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#warp-shuffle-functions) functions to implement them. Might be a bit faster.

Not necessary to change this, but worth keeping in mind in the future.

**Comment by ngimel** [comment]
And to make it cuda9-proof one has to use macros from THCDeviceUtils, such as https://github.com/pytorch/pytorch/blob/master/aten/src/THC/THCDeviceUtils.cuh#L66, rather than just __shfl (being deprecated in cuda 9) or __shfl_sync (not supported in cuda 8).

**Comment by zou3519** [comment]
Thanks for the suggestions! I'll take a look at the warp shuffles later and see how it goes (did not know that was a thing)
